server/src/main/java/org/elasticsearch/cluster/coordination/MasterHistory.java

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.coordination;
+
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterStateListener;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+/**
+ * This class represents a node's view of the history of which nodes have been elected master over the last 30 minutes. It is kept in
+ * memory, so when a node comes up it does not have any knowledge of previous master history before that point. This object is updated
+ * if and when the cluster state changes with a new master node.
+ */
+public class MasterHistory implements ClusterStateListener {
+    private volatile List<TimeAndMaster> masterHistory;
+    Supplier<Long> nowSupplier; // Can be changed for testing
+    /**
+     * The maximum amount of time that the master history covers.
+     */
+    public static final TimeValue MAX_HISTORY_AGE = new TimeValue(30, TimeUnit.MINUTES);
```

Comment:
Similarly here I think this should be configurable (with a static node-level setting). This is a sensible default IMO but time will tell if we're right about that.

Reply:
Is it OK for this to be a node-level setting instead of a cluster-level? We're passing around the history from different nodes, so one could be keeping 30 minutes and another 60.

Reply:
Yes that's fine IMO, it's not something that needs to be consistent across the cluster, in fact it's kind of a bug if it needs adjusting at all. Rather like [these settings](https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-discovery-settings.html#_expert_settings). If someone complains that they're getting weird behaviour from not setting it the same everywhere then I feel ok saying "¯\_(ツ)_/¯ don't do that"